pxr/base/tf/unicodeUtils.h

```diff
@@ -0,0 +1,625 @@
+//
+// Copyright 2023 Pixar
+//
+// Licensed under the Apache License, Version 2.0 (the "Apache License")
+// with the following modification; you may not use this file except in
+// compliance with the Apache License and the following modification to it:
+// Section 6. Trademarks. is deleted and replaced with:
+//
+// 6. Trademarks. This License does not grant permission to use the trade
+//    names, trademarks, service marks, or product names of the Licensor
+//    and its affiliates, except as required to comply with Section 4(c) of
+//    the License and to reproduce the content of the NOTICE file.
+//
+// You may obtain a copy of the Apache License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the Apache License with the above modification is
+// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the Apache License for the specific
+// language governing permissions and limitations under the Apache License.
+//
+#ifndef PXR_BASE_TF_UNICODE_UTILS_H
+#define PXR_BASE_TF_UNICODE_UTILS_H
+
+/// \file tf/unicodeUtils.h
+/// \ingroup group_tf_String
+
+#include "pxr/pxr.h"
+#include "pxr/base/tf/api.h"
+
+#include <string>
+#include <iterator>
+#include <stdexcept>
+#include <unordered_set>
+#include <unordered_map>
+#include <vector>
+#include <utility>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+/// \namespace TfUnicodeUtils
+/// \ingroup group_tf_String
+///
+/// Contains utility methods for processing classes of Unicode characters
+/// according to the Unicode standard.
+///
+/// Used by \ref page_tf_StringUtils to determine validity of identifiers and prim names.
+///
+namespace TfUnicodeUtils {
+
+    // 21-bits are required to store the maximum Unicode code point value
+    // so we use an unsigned integer to represent the value
+    extern TF_API std::unordered_set<uint32_t> xidStartClass;
+    extern TF_API std::unordered_set<uint32_t> xidContinueClass;
+    extern TF_API std::vector<std::pair<uint32_t, uint32_t>> xidStartRangeClass;
+    extern TF_API std::vector<std::pair<uint32_t, uint32_t>> xidContinueRangeClass;
+
+    void AppendUTF8Char(uint32_t codePoint, std::string& result);
+
+    /// Defines an iterator over a UTF-8 encoded string that extracts unicode code point values.
+    ///
+    /// UTF-8 is a variable length encoding, meaning that one Unicode character can be encoded in UTF-8 as
+    /// 1, 2, 3, or 4 bytes.  This iterator takes care of iterating the necessary characters in a string
+    /// and extracing the Unicode code point of each UTF-8 encoded character in the sequence.
+    ///
+    class utf8_const_iterator {
+
+    public:
+
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = uint32_t;
+        using difference_type = ptrdiff_t;
+        using pointer = uint32_t*;
```

Comment:
I suspect you're going to need to create a proxy type for `pointer` since `operator*` returns a temporary.  It also should probably be explicitly `const` since the type advertises itself as a `const_iterator`.

Reply:
see above